trieste/acquisition/function/function.py

```diff
@@ -583,6 +583,217 @@ def _update_expected_improvement_fn(
+                "The model must have a `reparam_sampler` method, which takes `num_samples` and "
+                "which returns a reparametrization `ReparametrizationSampler` to create smooth "
+                "samples at query points."
+            )
+
+        sampler = get_sampler(self._sample_size)
+
+        if not isinstance(sampler, ReparametrizationSampler):
+            raise ValueError(
+                "The sampler returned by the model's `reparam_sampler` method must be an instance "
+                "of a `trieste.models.interfaces.ReparametrizationSampler`."
+            )
+
+        if not isinstance(dataset, Dataset):
+            raise ValueError(
+                f"`dataset` must be a Dataset for {self.__repr__()}, received" f"{type(dataset)}"
+            )
+
+        tf.debugging.assert_positive(len(dataset))
+
+        samples_at_query_points = sampler.sample(dataset.query_points)
+        mean = tf.reduce_mean(samples_at_query_points, axis=0)
+
+        tf.debugging.assert_shapes(
+            [(mean, ["_", 1])], message="Expected model with output dimension [1]."
+        )
+
+        eta = tf.reduce_min(mean, axis=0)
+
+        return monte_carlo_expected_improvement(sampler, eta)
+
+
+class monte_carlo_expected_improvement(AcquisitionFunctionClass):
+    def __init__(self, sampler: ReparametrizationSampler[ProbabilisticModel], eta: TensorType):
+        r"""
+        Return a Monte Carlo based Expected Improvement (EI) acquisition function for
+        single-objective global optimization. Improvement is with respect to the current "best"
+        observation ``eta``, where an improvement moves towards the objective function's minimum
+        and the expectation is calculated with respect to the ``model`` posterior. For model
+        posterior :math:`f`, this is
+
+        .. math:: x \mapsto \mathbb E \left[ \max (\eta - f(x), 0) \right].
+
+        For the Monte Carlo version, the expectation is calculated by samples that we save.
+
+        :param sampler: The model sampler of the objective function.
+        :param eta: The "best" observation.
+        :return: The Monte Carlo expected improvement function. This function will raise
+            :exc:`ValueError` or :exc:`~tf.errors.InvalidArgumentError` if used with a batch size
+            greater than one.
+        """
+        self._sampler = sampler
+        self._eta = tf.Variable(eta)
+
+    @tf.function
+    def __call__(self, at: TensorType) -> TensorType:
+        tf.debugging.assert_shapes(
+            [(at, [..., 1, None])],
+            message="This acquisition function only supports batch sizes of one.",
+        )
+        samples = self._sampler.sample(tf.squeeze(at, -2))  # [S, N, 1]
+        improvement = tf.maximum(self._eta - samples, 0.0)  # [S, N, 1]
+        return tf.reduce_mean(improvement, axis=0)  # [N, 1]
+
+
+class MonteCarloAugmentedExpectedImprovement(
+    SingleModelAcquisitionBuilder[SupportsGetObservationNoise]
```

Comment:
we need here both HasReparamSampler and SupportsGetObservationNoise, the way to do that is to create an additional class above that defines this mixture:
```
@runtime_checkable
class SupportsReparamSamplerObservationNoise(
    HasReparamSampler, SupportsGetObservationNoise, Protocol
):
    """A model that supports both reparam_sampler and get_observation_noise."""

    pass
```

then we do 
```suggestion
    SingleModelAcquisitionBuilder[SupportsReparamSamplerObservationNoise]
```